README.md
# Patch manual
This repository has original files from the BLE Sniffer project by Nordic Semiconductor. [Link](https://www.nordicsemi.com/Software-and-tools/Development-Tools/nRF-Sniffer-for-Bluetooth-LE/) to the webpage. 
This project is really cool but it has only loadable HEXes for Nordic Semi development boards so you can't easily port it to every board with Nordic Semi MCU.

## Why I created this repo and put this project without any changes?
I use a BLE in my project so I try to get as more debug information as possible. And I decided to buy a powerful and cheap board for sniffing BLE traffic. This board isn't a development board from Nordic but _mostly_ it's compliant with the board [nRF52840 DK](https://www.nordicsemi.com/Software-and-tools/Development-Kits/nRF52840-DK) (PCA10056 in Nordic format). The main differences: 
- there is no onboard debugger;
- there is no reset pin/button (don't know why developers decided to do so);
- only two user buttons;
- cheap USB-UART bridge IC CP2102 by Silicon Labs;

So the last point is the reason why I created this patch. The original project default baudrate for UART is 1Mbaud but CP2102 IC maximum speed is 921600 baud. So It's impossible to get any useful data from this custom board because when Nordic python addon tries to open serial port with 1Mbaud the real speed will be 921600 baud and all data will be corrupted.
It's quite funny because API that is used between board and python script has a method that can change (reduce) baudrate but still you need to perform a first call somehow.

## What was changed in this project?
Almost nothing. I investigated that the firmware is used a UARTE0. This module has a register _BAUDRATE_ where firmware loads a new value. I didn't check a reference manual how to calculate this value (I think it's some sort of the divider/multiplier value for UART module clock) but SDK by Nordic has some precalculated values. So I understood that the value for 1Mbaud is _0x10000000_. With Segger J-Link debugger I found out that the value is copied from RAM with start address _0x20000240_ (+ 3 bytes because it's a 32bit value). But before loading from RAM somebody should init this value and this value must be in a flash memory, right? Yes and no. Yes it's loaded from flash address _0x5315_ but only the MSB value that is _0x10_ another three bytes are cleared with _0x00_. It's quite interesting because I think it's happened due to compiler optimization procedure. So we can change only MSB and thanks to all electronics gods because value _0x0F000000_ is compliant for BAUDRATE register and it equals to 921600 baud. So that's it. You need to change value at this address in the HEX file (the addresses are aligned so HEX _0x5315_ value is that you need) and recalculate hex check byte (it's just inversed sum of bytes for current line and we decrement byte by 1 the check byte will be incremented by 1). For this firmware version it'll be _0x80_ instead of _0x7F_ or you can check it with this [wiki link](https://en.wikipedia.org/wiki/Intel_HEX).
The last step that we should do is to change python script. You need to change script ___extcap/SnifferAPI/UART.py___ and add extra value to the list of values for _SNIFFER_BAUDRATES_. So the final value will be something like this:
```python
SNIFFER_BAUDRATES = [921600, 1000000, 460800]
```

Note. This step isn't mandatory because as I saw in Linux driver and also checked manually in Windows the driver setups a maximum baudrate for the CP2102 IC if the value is bigger than the maximum value that is 921600 baud.

## For which boards this manual is compliant?
First of all for boards that are compliant (at least UART pinout) with nRF52840 DK (PCA10056) and they have this cheap UART-USB bridge IC. This firmware is only compliant with nRF52840 MCU. 

### Information 
This manual is provided without any gurantee and author isn't responsible for steps you do.

### Extra
I've got a picture of my board by seller. 

![](board.jpg)

And also a link to a [seller](https://aliexpress.ru/item/32919832696.html).

release_notes.txt
Version 3.0.0:

This release switches the required Python runtime from 2.7 to 3.5 and above.
Make sure you have installed Python 3 in order to use this release.

New features:
    - All Python code has been ported to Python 3 and is no longer compatible
      with Python 2

Bug fixes:
    - Fix unstable behavior on Windows (coupled with Wireshark 3.0.7 and above)
    - Fix incorrect session key calculated when ENC_RSP and START_ENC_REQ was sent in same connection event.

Version 2.0.0:

This release adds support for following an LE secure encrypted connection when the Security Manager
of any of the devices has been placed in Debug mode. It also makes improvements to the UART implementation in
order to increase the throughput between the sniffer board and the host.

Deprecated features:
    - The version 2.0.0 sniffer firmware is not compatible with older host implementations due to a change in UART baudrate.

New features:
    - Support for following a connection using Security Manager defined debug public and private key
      pair, only one side (initiator or responder) needs to be set in Debug mode.
    - The firmware UART baudrate has been updated to 1M on all firmware versions.
    - The nRF52 UART implementation was moved to UARTE for increased efficiency and throughput.
    - Increased UART buffer size in the firmware leads to better tolerance for bursts of traffic in BLE connections.
    - Improved sniffer auto-detection should make Wireshark launch quicker and adds support for old and new firmwares.

Bug fixes:
    - Improved stability in finding connection request packets.
    - Fixed issue where sniffer was not able to decrypt packets when using Data Length Update.
    - Improved recovery when packets are lost in an encrypted connection.
    - Fixed a firmware bug where packets with a corrupted length field could lead to an invalid state
    - Fixed a host bug where the Sniffer would not be garbage collected due to a cyclic reference
    - Fixed a host bug where the PacketReader would wait for the full byte timeout before shutting down
    - Fixed a host bug where some packets would be discarded due to a read timeout
    - Added workraround for errata 102, 106 and 107 for all nRF52 devices


Version 2.0.0-beta-3:

Bugfix release to fix the the sniffing for DLE and 2Mbit PHY
    - Bugfix for Data Length Update procedure and 2Mbit PHY (PCA10040 only, requires Wireshark version 2.4.6 or
      newer)

Version 2.0.0-beta-2:

Summary:
This release updates to use version 2 of the sniffer UART protocol which supports UART packets
larger than 256 bytes.

New features:
    - Support for v2 of sniffer UART protocol. SnifferAPI supports firmware v1 sniffer UART
      protocol.
    - Support for Data Length Update procedure (PCA10040 only, requires Wireshark version 2.4.6 or
      newer)

Bug fixes:
    - Improved UART stability.

Version 2.0.0-b1:

Summary:
This release replaces the old nRF Sniffer.exe control window used for sending
commands to the sniffer hardware. This functionality is now moved to Wireshark.

As a result of this, nrf_sniffer now supports all platforms supported by
Wireshark (Windows, Linux, OSX).

The supported hardware has also been updated to PCA10028, PCA10031 and PCA10040.
Note that PCA10040 uses nRF52 while the others uses nRF51.

Deprecated features:
    - nRF_Sniffer.exe application (replaced with support in Wireshark)

New features:
    - Support for nRF52 (PCA10040).
    - Support for 2mbit mode (PCA10040 only, Requires Wireshark version 2.4.2 or newer).

Known issues:
    - Several lost packets is reported when starting the sniffer.


extcap/SnifferAPI/UART.py
# Copyright (c) 2017, Nordic Semiconductor ASA
# 
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 
#    1. Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
# 
#    2. Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
# 
#    3. Neither the name of Nordic Semiconductor ASA nor the names of
#       its contributors may be used to endorse or promote products
#       derived from this software without specific prior written
#       permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY, AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NORDIC
# SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
# OF THE POSSIBILITY OF SUCH DAMAGE.

import collections
import logging
import serial
from threading import Thread, Event

import serial.tools.list_ports as list_ports

from . import Exceptions
from . import Packet


SNIFFER_OLD_DEFAULT_BAUDRATE = 460800
# Baudrates that should be tried (add more if required)
SNIFFER_BAUDRATES = [921600, 1000000, 460800]


def find_sniffer(write_data=False):
    open_ports = list_ports.comports()

    sniffers = []
    for port in [x.device for x in open_ports]:
        for rate in SNIFFER_BAUDRATES:
            reader = None
            try:
                reader = Packet.PacketReader(portnum=port, baudrate=rate)
                try:
                    if write_data:
                        reader.sendPingReq()
                        _ = reader.decodeFromSLIP(0.1, complete_timeout=0.1)
                    else:
                        _ = reader.decodeFromSLIP(0.3, complete_timeout=0.3)

                    # FIXME: Should add the baud rate here, but that will be a breaking change
                    sniffers.append(port)
                    break
                except (Exceptions.SnifferTimeout, Exceptions.UARTPacketError):
                    pass
            except (serial.SerialException, ValueError):
                continue
            finally:
                if reader is not None:
                    reader.doExit()
    return sniffers


def find_sniffer_baudrates(port):
    for rate in SNIFFER_BAUDRATES:
        reader = None
        try:
            reader = Packet.PacketReader(portnum=port, baudrate=rate)
            try:
                reader.sendPingReq()
                _ = reader.decodeFromSLIP(0.1, complete_timeout=0.1)
                # TODO: possibly include additional rates based on protocol version
                return {"default": rate, "other": []}
            except (Exceptions.SnifferTimeout, Exceptions.UARTPacketError):
                pass
        finally:
            if reader is not None:
                reader.doExit()
    return None


class Uart:
    def __init__(self, portnum=None, baudrate=None):
        self.ser = None
        try:
            if baudrate is not None and baudrate not in SNIFFER_BAUDRATES:
                raise Exception("Invalid baudrate: " + str(baudrate))

            logging.info('Opening serial port {}'.format(portnum))
            self.ser = serial.Serial(
                port=portnum,
                baudrate=9600,
                rtscts=True
            )
            self.ser.baudrate = baudrate

        except Exception as e:
            if self.ser:
                self.ser.close()
                self.ser = None
            raise

        self.read_queue = collections.deque()
        self.read_queue_has_data = Event()

        self.worker_thread = Thread(target=self._read_worker)
        self.reading = True
        self.worker_thread.setDaemon(True)
        self.worker_thread.start()

    def _read_worker(self):
        self.ser.reset_input_buffer()
        while self.reading:
            try:
                # Read any data available, or wait for at least one byte
                data_read = self.ser.read(self.ser.in_waiting or 1)
                #logging.info('type: {}'.format(data_read.__class__))
                self._read_queue_extend(data_read)
            except serial.SerialException as e:
                logging.info("Unable to read UART: %s" % e)
                self.reading = False
                return

    def close(self):
        if self.ser:
            logging.info("closing UART")
            self.reading = False
            # Wake any threads waiting on the queue
            self.read_queue_has_data.set()
            if hasattr(self.ser, "cancel_read"):
                self.ser.cancel_read()
                self.worker_thread.join()
                self.ser.close()
            else:
                self.ser.close()
                self.worker_thread.join()
            self.ser = None

    def __del__(self):
        self.close()

    def switchBaudRate(self, newBaudRate):
        self.ser.baudrate = newBaudRate

    def readByte(self, timeout=None):
        r = self._read_queue_get(timeout)
        #logging.info('rbtype: {}'.format(r.__class__))
        #logging.info('r: {}'.format(r))
        return r
        
    def writeList(self, array):
        try:
            #logging.info('array: {}'.format(array))
            self.ser.write(array)
        except serial.SerialTimeoutException:
            logging.info("Got write timeout, ignoring error")

        except serial.SerialException as e:
            self.ser.close()
            raise e

    def _read_queue_extend(self, data):
        if len(data) > 0:
            self.read_queue.extend(data)
            self.read_queue_has_data.set()

    def _read_queue_get(self, timeout=None):
        data = None
        if self.read_queue_has_data.wait(timeout):
            self.read_queue_has_data.clear()
            try:
                data = self.read_queue.popleft()
            except IndexError:
                # This will happen when the class is destroyed
                return None
            if len(self.read_queue) > 0:
                self.read_queue_has_data.set()
        return data


def list_serial_ports():
    # Scan for available ports.
    return list_ports.comports()

if __name__ == "__main__":
    import time
    t_start = time.time()
    s = find_sniffer()
    tn = time.time()
    print(s)
    print("find_sniffer took %f seconds" % (tn - t_start))
    for p in s:
        t = time.time()
        print(find_sniffer_baudrates(p))
        tn = time.time()
        print("find_sniffer_baudrate took %f seconds" % (tn - t))
    tn = time.time()
    print("total runtime %f" % (tn - t_start))
